libs/langchain-cloudflare/src/chat_models.ts

```diff
@@ -3,25 +3,109 @@ import {
   SimpleChatModel,
   type BaseChatModelParams,
 } from "@langchain/core/language_models/chat_models";
-import type { BaseLanguageModelCallOptions } from "@langchain/core/language_models/base";
+import type {
+  BaseLanguageModelCallOptions,
+  BaseLanguageModelInput,
+  ToolDefinition,
+} from "@langchain/core/language_models/base";
 import {
+  AIMessage,
   AIMessageChunk,
   BaseMessage,
   ChatMessage,
 } from "@langchain/core/messages";
-import { ChatGenerationChunk } from "@langchain/core/outputs";
+import { ChatGenerationChunk, ChatResult } from "@langchain/core/outputs";
 import { getEnvironmentVariable } from "@langchain/core/utils/env";
 import { CallbackManagerForLLMRun } from "@langchain/core/callbacks/manager";
 
-import type { CloudflareWorkersAIInput } from "./llms.js";
+import { StructuredToolInterface } from "@langchain/core/tools";
+import { Runnable } from "@langchain/core/runnables";
+import { isStructuredTool } from "@langchain/core/utils/function_calling";
+import { zodToJsonSchema } from "zod-to-json-schema";
+import { ToolCall } from "@langchain/core/messages/tool";
 import { convertEventStreamToIterableReadableDataStream } from "./utils/event_source_parse.js";
+import type { CloudflareWorkersAIInput } from "./llms.js";
+
+type CloudflareTool = {
+  name: string;
+  description: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  parameters: Record<string, any>;
+};
+
+type CloudflareToolResponse = {
+  name: string;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  arguments: Record<string, any>;
+};
+
+type Tool =
+  | Record<string, unknown>
+  | CloudflareTool
+  | StructuredToolInterface
+  | ToolDefinition;
+
+const convertCloudflareToolsToLangChainTools = (
+  tools: CloudflareToolResponse[]
+): ToolCall[] =>
+  tools.map((tc) => ({
+    name: tc.name,
+    args: tc.arguments,
+  }));
+
+function convertToCloudflareTools(tools: Tool[]): CloudflareTool[] {
+  if (tools.every(isStructuredTool)) {
+    return (tools as StructuredToolInterface[]).map((tc) => ({
+      name: tc.name,
+      description: tc.description,
+      parameters: zodToJsonSchema(tc.schema),
+    }));
+  }
+  if (
+    tools.every((t) =>
+      Boolean(
+        "type" in t &&
+          t.type === "function" &&
+          "function" in t &&
+          typeof t.function === "object"
+      )
+    )
+  ) {
+    return (tools as ToolDefinition[]).map((tc) => ({
+      name: tc.function.name,
+      description:
+        tc.function.description ?? `A function named ${tc.function.name}.`,
+      parameters: tc.function.parameters,
+    }));
+  }
+  if (
+    tools.every((t) =>
+      Boolean(
```

Comment:
This constructor has some funny edge cases IIRC - let's just use `!!` to get truthy values only if we must